clients/drcachesim/tracer/tracer.cpp

```diff
@@ -608,23 +609,196 @@ clean_call(void)
 }
 
 /***************************************************************************
- * Tracing instrumentation.
+ * Alternating tracing-no-tracing feature.
  */
 
-static int tracing_enabled;
+/* We have two modes: count instructions only, and full trace. */
+enum {
+    BBDUP_MODE_TRACE = 0,
+    BBDUP_MODE_COUNT = 1,
+};
+
+#if defined(X86_64) || defined(AARCH64)
+#    define DELAYED_CHECK_INLINED 1
+#else
+/* XXX we don't have the inlining implemented yet for 32-bit architectures. */
+#endif
+
+/* This holds one of the BBDUP_MODE_ enum values, but drbbdup requires that it
+ * be pointer-sized.
+ */
+static std::atomic<ptr_int_t> tracing_disabled;
+
+static dr_emit_flags_t
+event_bb_analysis(void *drcontext, void *tag, instrlist_t *bb, bool for_trace,
```

Comment:
Agreed that tracer.cpp is getting large and unwieldy.  My hesitation is that I have a large branch for the #3995 trace windows based on top of this one for which a refactor here would cause a serious merge headache...what if after both branches are in I then did a refactor, splitting up the bbdup and possibly other things?

Reply:
Sure, makes sense to have it as a separate PR anyways!